Review

Soulgado opened the pull request:
First iteration

Soulgado committed:
feat(component): added logic for showing sidebar

Soulgado committed:
feat(component): added ui for side form

Soulgado committed:
feat(component): added logic for submitting form

Soulgado committed:
feat(component): sidebar is closed on click on document, except sidebar and arrow

gennady-bars commented on src/index.tsx:
```diff
@@ -13,19 +13,20 @@ const domNode = document.getElementById('root') as HTMLDivElement;
+	const [styleOptions, setStyleOptions] = useState(defaultArticleState);
 	return (
 		<main
 			className={clsx(styles.main)}
```
**Можно лучше**

* Лучше выделить компонент  `App` в отдельный файл, чтобы не смешивать все в `index.js`
* `clsx` можно не использовать, если только **один** селектор передается

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,135 @@
+		setFontSizeOption(fontSizeOptions[0]);
+		setFontColor(fontColors[0]);
+		setBackgroundColor(backgroundColors[0]);
+		setContentWidth(contentWidthArr[0]);
```
* Начальные данные прописаны уже в объекте `defaultArticleState` в файле `articleProps.ts`. Вот этот объект нужно использовать  для дефолтных значений, а не опять собирать через индексы из разных объектов

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,135 @@
+		});
+	}
+
+	useEffect(() => {
```
* Нужно **останавливать эффект**, если форма закрыта, чтобы не навешивать обработчик 


```js

	useEffect(() => {
		if (!menuIsOpen) return;  // останавливаем, если закрыто
		const handleCloseEvent = () => {
```

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,135 @@
+		const handleCloseEvent = () => {
+			setMenuIsOpen(false);
+		};
+		document.addEventListener('click', handleCloseEvent);
+
+		return () => {
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для правильного навешивания и удаления обработчиков `Escape` и оверлея (клика снаружи блока):

```js
import { useEffect } from "react";

type TUseClose = {
  isOpen: boolean;
  onClose: () => void;
  rootRef: React.RefObject<HTMLElement>;
};
// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export function useClose({isOpen, onClose, rootRef}: TUseClose) {
  useEffect(() => {
    if (!isOpen) return; // останавливаем действие эффекта, если закрыто

   
    function handleClickOutside(event: MouseEvent) {
      const { target } = event;
      const isOutsideClick =
        target instanceof Node && // проверяем, что это `DOM`-элемент
        rootRef.current &&
        !rootRef.current.contains(target);  // проверяем, что кликнули на элемент, который находится не внутри нашего блока
      if (isOutsideClick) {
        onClose();
      }
    }

    const handleEscape = (e: KeyboardEvent) => {
      if (e.key === "Escape") {
        onClose();
      }
    };

    document.addEventListener("keydown", handleEscape);
    document.addEventListener("mousedown", handleClickOutside);

    //  обязательно удаляем обработчики в `clean-up`- функции
    return () => {
      document.removeEventListener("keydown", handleEscape);
      document.removeEventListener("mousedown", handleClickOutside);
    };
    // обязательно следим за `isOpen`, чтобы срабатывало только при открытии, а не при любой перерисовке компонента
  }, [isOpen, onClose, rootRef]);
}


```

Этот код помещают в отдельный файл `useClose.ts` в папке `hooks` и импортируют хук туда, где нужно.

И теперь можно использовать в любом компоненте (попапе, меню и тд), которому нужно установить эти обработчики. 

Пример того, как это использовать:
 
```js
    // создаем `ref` для нашей формы и указываем его на внешнем блоке с формой
    const formRef = useRef<HTMLElement>(null);
    
    useClose({
        isOpen: isMenuOpen,
        onClose: closeMenu,
        rootRef: formRef,
    });

```

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,135 @@
+					(event) =>
+						event.stopPropagation() /* so sidebar isn't closed on click on it */
+				}
+				className={clsx(styles.container, menuIsOpen && styles.container_open)}>
```
**Можно лучше**

 
* Библиотека `clsx` имеет возможность добавления стилей по **условию**: если оно будет `true`, то добавится селектор, а если `false`, то нет. Подробнее тут  https://habr.com/ru/articles/649381/

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,135 @@
+						event.stopPropagation() /* so sidebar isn't closed on click on it */
+				}
+				className={clsx(styles.container, menuIsOpen && styles.container_open)}>
+				<form className={styles.form} onSubmit={handleSubmit}>
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:

* Хорошо реализовали форму из предоставленных компонентов
* Хорошая логика передачи данных от `App` в форму и обратно
* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. 
* Отлично, что нет типов `any`
* Отлично типизировали проект


но есть некоторые недочеты:

* Начальные данные прописаны уже в объекте `defaultArticleState` в файле `articleProps.ts`. Вот этот объект нужно использовать  для дефолтных значений, а не опять собирать через индексы из разных объектов
* Нужно **останавливать эффект**, если форма закрыта, чтобы не навешивать обработчик 

**Можно лучше**

* Лучше выделить компонент  `App` в отдельный файл, чтобы не смешивать все в `index.js`
* `clsx` можно не использовать, если только **один** селектор передается
* Библиотека `clsx` имеет возможность добавления стилей по **условию**: если оно будет `true`, то добавится селектор, а если `false`, то нет. Подробнее тут  https://habr.com/ru/articles/649381/

Soulgado committed:
refactor(component): refactor minor errors

Default values for inputs are taken from DefaultArticleState, added condition, so close event is not
attached to document if sidebar is closed

Soulgado committed:
style(component): changed conditional class addition in ArticleParamsForm component

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
